src/client.ts

```diff
@@ -254,7 +261,32 @@ export class Client {
     //   to simply do `const queryInterpolation: TaggedTypeFormat.encode(query)`
     const queryInterpolation = query.toQuery(options).query;
 
-    return this.#query(queryInterpolation, options);
+    return this.#queryWithRetries(queryInterpolation, options);
+  }
+
+  async #queryWithRetries<T extends QueryValue>(
+    queryInterpolation: string | QueryInterpolation,
+    options?: QueryOptions,
+    attempt = 0
+  ): Promise<QuerySuccess<T>> {
+    const maxBackoff =
+      this.clientConfiguration.max_backoff ?? DEFAULT_CLIENT_CONFIG.max_backoff;
+    const maxAttempts =
+      this.clientConfiguration.max_attempts ??
+      DEFAULT_CLIENT_CONFIG.max_attempts;
+    const backoffMs =
+      Math.min(Math.random() * 2 ** attempt, maxBackoff) * 1_000;
+
+    const wait = (ms: number) => new Promise((r) => setTimeout(r, ms));
+    attempt += 1;
```

Comment:
Is a random number preferred here? That could mean any retry could be near 0s.

We could ramp up the exponential backoff as a function ramping up to equal `max_backoff` at `max_attempts`

```suggestion
    // backoff formula:  t = e ^ (r/c)
    // where t is the backoff time, r is retry attempt, c is a constant
    // t_max = e ^ (r_max/c)  =>  c = r_max/ln(t_max)
    
    attempt += 1;
    const c = (maxAttempts - 1) / Math.log(maxBackoff)
    const backoffMs = Math.exp(attempt/c) * 1_000;

    const wait = (ms: number) => new Promise((r) => setTimeout(r, ms));
```

Reply:
Yeah, this is the impl we've chosen in all drivers.

Reply:
okie dokie then.